### PR TITLE
[Synfig Studio] Allow to set number of rendering threads from Preferences dialog

### PR DESCRIPTION
**Pull request by SushantAA**
fix - #1410 
I am new to GTK :turtle: 
pls help me in verifying the logic 
currently, it only changes the core to 3 ( I don't know how to check [ how many cores is running ] )

### CODE REVIEW

File: synfig-core/src/synfig/threadpool.h [hunk 1]
```diff
@@ -98,6 +99,8 @@ class ThreadPool {
 	void enqueue(const Slot &slot);
 	void wait(std::condition_variable &cond, std::unique_lock<std::mutex>& lock);
 
+	void set_threads(int no_core);
```
Comment (rodolforg) [inline synfig-core/src/synfig/threadpool.h:102]: Please rename the method to `set_num_threads` ('num' from 'number'): let the method name descriptive in order to help code reading.

And rename `no_core` to `num_cores`or something like that: 'no' means 'without' here

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 2]
```diff
@@ -535,7 +542,7 @@ Dialog_Setup::create_render_page(PageInfo pi)
 	preview_background_color_button.signal_color_set().connect(
 		sigc::mem_fun(*this, &studio::Dialog_Setup::on_preview_background_color_changed) );
 
-}
+}
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:545]: trim whitespace

File: synfig-studio/src/gui/app.h [hunk 3]
```diff
@@ -51,6 +51,7 @@
 #include <synfig/layers/layer_bitmap.h>
 #include <synfig/string.h>
 #include <synfig/time.h>
+#include <synfig/threadpool.h>
```
Comment (rodolforg) [inline synfig-studio/src/gui/app.h:54]: Why did you add it?
Neither `app.h` nor `app.cpp` need it.

File: synfig-studio/src/gui/app.cpp [hunk 4]
```diff
@@ -285,6 +285,7 @@ std::set< String > studio::App::brushes_path;
 String studio::App::image_editor_path;
 
 String studio::App::sequence_separator(".");
+String studio::App::number_of_cores("" + ('0'+g_get_num_processors()) );
```
Comment (rodolforg) [inline synfig-studio/src/gui/app.cpp:288]: use standard c++ function : `std::to_string()` instead of ASCII tricks.
But why you store an integer number as text?!

File: synfig-core/src/synfig/threadpool.h [hunk 5]
```diff
@@ -43,6 +43,7 @@
 
 /* === C L A S S E S & S T R U C T S ======================================= */
 
+using namespace synfig;
```
Comment (rodolforg) [inline synfig-core/src/synfig/threadpool.h:46]: [Never use `using namespace` in a header file](https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers).

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 6]
```diff
@@ -669,6 +676,8 @@ Dialog_Setup::on_restore_pressed()
 		fps_template_combo->set_active_text(DEFAULT_PREDEFINED_FPS);
 		adj_pref_fps->set_value(24.0);
 		image_sequence_separator.set_text(".");
+		core_render.set_text("" + ('0'+g_get_num_processors()));
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:679]: as already said in this review: use `std::to_string()`

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 7]
```diff
@@ -904,6 +916,12 @@ Dialog_Setup::on_autobackup_changed()
 	auto_backup_interval.set_sensitive(toggle_autobackup.get_active());
 }
 
+void
+Dialog_Setup::on_core_change()
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:920]: give a better name to method: the 'core' didn't change. ;)

File: synfig-core/src/synfig/threadpool.cpp [hunk 8]
```diff
@@ -275,6 +275,21 @@ ThreadPool::enqueue(const Slot &slot) {
...
+	max_running_threads = g_get_num_processors();
+	if(no_core!=0){
+		max_running_threads = no_core;
+	}
+	if (const char *s = getenv("SYNFIG_GENERIC_THREADS"))
+		max_running_threads = atoi(s) + 1;
+
+	if (max_running_threads < 2) max_running_threads = 2;
+	if (max_running_threads > 2) --max_running_threads;
+	++running_threads;
```
Comment (rodolforg) [inline synfig-core/src/synfig/threadpool.cpp:290]: I couldn't understand why did you do it…

File: synfig-core/src/synfig/threadpool.cpp [hunk 9]
```diff
@@ -275,6 +275,21 @@ ThreadPool::enqueue(const Slot &slot) {
...
+}
+
```
Comment (rodolforg) [inline synfig-core/src/synfig/threadpool.cpp:292]: Maybe the `max_running_threads`should be bounded to num_processors, even if 'no_core' is greater than it.

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 10]
```diff
@@ -501,6 +501,13 @@ Dialog_Setup::create_render_page(PageInfo pi)
 	pi.grid->attach(image_sequence_separator, 1, row, 1, 1);
 	image_sequence_separator.set_hexpand(true);
 
+	// render - cores
+	attach_label(pi.grid, _("Limit render to number of cores"), ++row);
+	pi.grid->attach(core_render, 1, row, 1, 1);
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:506]: `core_render` is really a bad name choice :P It doesn't mean what it stores/is.

Comment (SushantAA) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp:506]: I know they are bad names [ Sorry I forgot to write it :disappointed:   ] but I just wanted to verify logic. because I am new here.
Sorry for the trouble in pointing out the bad name.

File: synfig-studio/src/gui/app.h [hunk 11]
```diff
@@ -227,6 +228,7 @@ class App : public Gtk::Main, private IconController
 	static synfig::String predefined_fps;
 	static float preferred_fps;
 	static synfig::String sequence_separator;
+	static synfig::String number_of_cores;
```
Comment (ice0) [inline synfig-studio/src/gui/app.h:231]: Better to use integer value here

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp (hunk 10 above)
Comment (ice0) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp:506]: No problem here. Logic is looks correct. Also you need to check CI builds. Currently they fails with error.

Comment (rodolforg) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp:506]: @SushantAA I'm sorry if it sounded harsh, it wasn't my intention. It was my last comment about variable name and I explained it better in previous comments, so I thought I could be more concise. I was wrong :(

Comment (SushantAA) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp:506]: @rodolforg nothing was harsh, it's fine  :smiley:

File: synfig-studio/src/gui/app.h (hunk 3 above)
Comment (SushantAA) [reply synfig-studio/src/gui/app.h:54]: to change the number of cores by calling a function 
https://github.com/SushantAA/synfig/blob/issue-1410/synfig-core/src/synfig/threadpool.cpp#L280

File: synfig-studio/src/gui/dialogs/dialog_setup.h [hunk 12]
```diff
@@ -163,6 +164,7 @@ class Dialog_Setup : public Dialog_Template
 	Gtk::Entry        image_sequence_separator;
 	Gtk::ComboBoxText workarea_renderer_combo;
 	Gtk::Switch       toggle_play_sound_on_render_done;
+	Gtk::Entry		  number_of_core_select;
```
Comment (SushantAA) [inline synfig-studio/src/gui/dialogs/dialog_setup.h:167]: ah, i will correct that space

**Comment by SushantAA** [comment]
Basically pls 
* verify the logic and the implementation
* I have no idea how to check the distribution of core ( I use OS: Linux mint 20)
* I am not able to set the initial display of the number of cores on the Gtk::Entry

File: synfig-studio/src/gui/app.cpp [hunk 13]
```diff
@@ -285,6 +285,7 @@ std::set< String > studio::App::brushes_path;
 String studio::App::image_editor_path;
 
 String studio::App::sequence_separator(".");
+int studio::App::number_of_cores(g_get_num_processors());
```
Comment (rodolforg) [inline synfig-studio/src/gui/app.cpp:288]: cosmetic: it isn't related to surrounding variables. Move it to below ` String studio::App::workarea_renderer; ` separated by a blank line :)

**Review by rodolforg — CHANGES_REQUESTED** [review]
As ice0 said, your logic seems ok :)

**Comment by ice0** [comment]
> * verify the logic and the implementation

Looks good. 

> * I have no idea how to check the distribution of core ( I use OS: Linux mint 20)

Can you explain this with more details?

> * I am not able to set the initial display of the number of cores on the Gtk::Entry

You've already done it. But there is a bug in your code. Please check `~/.config/synfig/settings-1.4` to see why it doesn't work.

File: synfig-studio/src/gui/app.h (hunk 3 above)
Comment (SushantAA) [reply synfig-studio/src/gui/app.h:54]: I tried removing it but it showed a build error

Comment (rodolforg) [reply synfig-studio/src/gui/app.h:54]: You should read the build error to check what file misses this header and fix it properly.

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 14]
```diff
@@ -1064,6 +1080,9 @@ Dialog_Setup::refresh()
 	// Refresh the sequence separator
 	image_sequence_separator.set_text(App::sequence_separator);
 
+	// Refresh the number of core
+	number_of_core_select->set_value(int(adj_number_of_core->get_value()));
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:1084]: This isn't right: `refresh()` is to update the form fields, and not set the field with its "own" value.
You could keep the App "member" variable you created commits ago, but now removed and used it here :) Something like:
`number_of_core_select->set_value(App::num_threads);`

File: synfig-studio/src/gui/dialogs/dialog_setup.h [hunk 15]
```diff
@@ -163,6 +164,8 @@ class Dialog_Setup : public Dialog_Template
 	Gtk::Entry        image_sequence_separator;
 	Gtk::ComboBoxText workarea_renderer_combo;
 	Gtk::Switch       toggle_play_sound_on_render_done;
+	Glib::RefPtr<Gtk::Adjustment> adj_number_of_core;
+	Gtk::SpinButton*	number_of_core_select;
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.h:168]: Answering your comment here: you can align line 168 with lines 164-166, but with spaces, not tab.

File: synfig-studio/src/gui/dialogs/dialog_setup.h [hunk 16]
```diff
@@ -99,6 +99,7 @@ class Dialog_Setup : public Dialog_Template
 	void on_def_background_type_changed(); //bound on clicked
 	void on_def_background_color_changed();
 	void on_def_background_image_set();
+	void on_number_of_core_select();
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.h:102]: Looking to other signals slots/callbacks name: prefer `on_number_of_core_changed()`.
In earlier commits you used `on_core_changed()`, and it led to a confusion: did the core changed? No, the number of core (or threads).

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 17]
```diff
@@ -904,6 +914,12 @@ Dialog_Setup::on_autobackup_changed()
 	auto_backup_interval.set_sensitive(toggle_autobackup.get_active());
 }
 
+void
+Dialog_Setup::on_number_of_core_select()
+{
+	ThreadPool::instance().set_num_threads(int(adj_number_of_core->get_value()));
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:920]: Gtk/gtkmm has a method here that help you: `Gtk::SpinButton::get_value_as_int()` :D

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp [hunk 18]
```diff
@@ -76,6 +77,7 @@ Dialog_Setup::Dialog_Setup(Gtk::Window& parent):
 	adj_pref_x_size(Gtk::Adjustment::create(480,1,10000,1,10,0)),
 	adj_pref_y_size(Gtk::Adjustment::create(270,1,10000,1,10,0)),
 	adj_pref_fps(Gtk::Adjustment::create(24.0,1.0,100,0.1,1,0)),
+	adj_number_of_core(Gtk::Adjustment::create(g_get_num_processors(),1,g_get_num_processors(),1,10,0)),
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/dialog_setup.cpp:80]: Reading the code of `ThreadPool::set_num_threads(int num_cores)`, the minimum value should be 2, not 1.

**Review by rodolforg — CHANGES_REQUESTED** [review]
@ice0 do you think it should use the word `core` instead of `thread`?

File: synfig-studio/src/gui/dialogs/dialog_setup.h (hunk 16 above)
Comment (SushantAA) [reply synfig-studio/src/gui/dialogs/dialog_setup.h:102]: is `on_number_of_thread_changed()`  fine

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp (hunk 17 above)
Comment (SushantAA) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp:920]: get_value_as_int() shows error as it is not a member of Gtk::Adjustment
in other places also it is not used
https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/dialogs/dialog_setup.cpp#L760

File: synfig-studio/src/gui/dialogs/dialog_setup.h (hunk 15 above)
Comment (SushantAA) [reply synfig-studio/src/gui/dialogs/dialog_setup.h]: wrong_align
I don't know why it's happening
https://github.com/SushantAA/synfig/blob/issue-1410/synfig-studio/src/gui/dialogs/dialog_setup.h#L168

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp (hunk 17 above)
Comment (rodolforg) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp]: It's for `Gtk::SpinButton`, not for its `Gtk::Adjustment`. But ok.

File: synfig-studio/src/gui/dialogs/dialog_setup.h (hunk 16 above)
Comment (ice0) [reply synfig-studio/src/gui/dialogs/dialog_setup.h]: Agree. But better on_number_of_thread*s*_changed()

File: synfig-studio/src/gui/dialogs/dialog_setup.cpp (hunk 14 above)
Comment (ice0) [reply synfig-studio/src/gui/dialogs/dialog_setup.cpp]: @rodolforg I think threads is better. We can't change CPU cores :)

**Comment by ice0** [comment]
Looks fine. The only suggestion is to rename cores to threads, because we changing number of rendering threads.

**Comment by SushantAA** [comment]
@rodolforg @ice0 I think its complete

**Review by ice0 — APPROVED** [review]
Looks good to me

**Review by rodolforg — APPROVED** [review]

**Comment by ice0** [comment]
Merged. Thank you!

File: synfig-studio/src/gui/app.cpp [hunk 19]
```diff
@@ -542,6 +543,11 @@ class Preferences : public synfigapp::Settings
 				value=App::sequence_separator;
 				return true;
 			}
+			if(key=="number_of_threads")
+			{
+				value=App::number_of_threads;
```
Comment (rodolforg) [inline synfig-studio/src/gui/app.cpp:548]: @SushantAA  maybe you should convert this integer into string before assign to value?

Comment (SushantAA) [reply synfig-studio/src/gui/app.cpp:548]: It should be a string :grimacing:
@rodolforg should the initial setting of threads should be 2 or 12. currently, it's 2 but I think 12 will be better

Comment (rodolforg) [reply synfig-studio/src/gui/app.cpp:548]: Look few lines above: use `etl::strprintf()` to 'convert' the `App::number_of_threads`.

IIRC, you make it to be the maximum threads as default.

Comment (SushantAA) [reply synfig-studio/src/gui/app.cpp:548]: https://github.com/synfig/synfig/pull/2023
